File: platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/maps/InfoWindowManager.java
package org.maplibre.android.maps;

import android.text.TextUtils;

import androidx.annotation.Nullable;

import org.maplibre.android.annotations.InfoWindow;
import org.maplibre.android.annotations.Marker;

import java.util.ArrayList;
import java.util.List;

/**
 * Responsible for managing InfoWindows shown on the Map.
 * <p>
 * Maintains a {@link List} of opened {@link InfoWindow} and tracks configurations as
 * allowConcurrentMultipleInfoWindows which allows to have multiple {@link InfoWindow} open at the
 * same time. Responsible for managing listeners as
 * {@link MapLibreMap.OnInfoWindowClickListener} and
 * {@link MapLibreMap.OnInfoWindowLongClickListener}.
 * </p>
 */
class InfoWindowManager {

  private final List<InfoWindow> infoWindows = new ArrayList<>();

  @Nullable
  private MapLibreMap.InfoWindowAdapter infoWindowAdapter;
  private boolean allowConcurrentMultipleInfoWindows;

  @Nullable
  private MapLibreMap.OnInfoWindowClickListener onInfoWindowClickListener;
  @Nullable
  private MapLibreMap.OnInfoWindowLongClickListener onInfoWindowLongClickListener;
  @Nullable
  private MapLibreMap.OnInfoWindowCloseListener onInfoWindowCloseListener;

  void update() {
    if (!infoWindows.isEmpty()) {
      for (InfoWindow infoWindow : infoWindows) {
        infoWindow.update();
      }
    }
  }

  void setInfoWindowAdapter(@Nullable MapLibreMap.InfoWindowAdapter infoWindowAdapter) {
    this.infoWindowAdapter = infoWindowAdapter;
  }

  @Nullable
  MapLibreMap.InfoWindowAdapter getInfoWindowAdapter() {
    return infoWindowAdapter;
  }

  void setAllowConcurrentMultipleOpenInfoWindows(boolean allow) {
    allowConcurrentMultipleInfoWindows = allow;
  }

  boolean isAllowConcurrentMultipleOpenInfoWindows() {
    return allowConcurrentMultipleInfoWindows;
  }

  boolean isInfoWindowValidForMarker(@Nullable Marker marker) {
    return marker != null && (!TextUtils.isEmpty(marker.getTitle()) || !TextUtils.isEmpty(marker.getSnippet()));
  }

  void setOnInfoWindowClickListener(@Nullable MapLibreMap.OnInfoWindowClickListener listener) {
    onInfoWindowClickListener = listener;
  }

  @Nullable
  MapLibreMap.OnInfoWindowClickListener getOnInfoWindowClickListener() {
    return onInfoWindowClickListener;
  }

  void setOnInfoWindowLongClickListener(@Nullable MapLibreMap.OnInfoWindowLongClickListener listener) {
    onInfoWindowLongClickListener = listener;
  }

  @Nullable
  MapLibreMap.OnInfoWindowLongClickListener getOnInfoWindowLongClickListener() {
    return onInfoWindowLongClickListener;
  }

  void setOnInfoWindowCloseListener(@Nullable MapLibreMap.OnInfoWindowCloseListener listener) {
    onInfoWindowCloseListener = listener;
  }

  @Nullable
  MapLibreMap.OnInfoWindowCloseListener getOnInfoWindowCloseListener() {
    return onInfoWindowCloseListener;
  }

  public void add(InfoWindow infoWindow) {
    infoWindows.add(infoWindow);
  }
}

File: platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/maps/InfoWindowManager.kt
package org.maplibre.android.maps

import org.maplibre.android.annotations.InfoWindow
import org.maplibre.android.annotations.Marker

/**
 * Responsible for managing InfoWindows shown on the Map.
 *
 * Maintains a [List] of opened [InfoWindow] and tracks configurations as
 * allowConcurrentMultipleInfoWindows which allows to have multiple [InfoWindow] open at the
 * same time. Responsible for managing listeners as
 * [MapLibreMap.OnInfoWindowClickListener] and
 * [MapLibreMap.OnInfoWindowLongClickListener].
 */
internal class InfoWindowManager {
    private val infoWindows: MutableList<InfoWindow> = ArrayList()

    private var infoWindowAdapter: MapLibreMap.InfoWindowAdapter? = null
    private var allowConcurrentMultipleInfoWindows = false

    private var onInfoWindowClickListener: MapLibreMap.OnInfoWindowClickListener? = null
    private var onInfoWindowLongClickListener: MapLibreMap.OnInfoWindowLongClickListener? = null
    private var onInfoWindowCloseListener: MapLibreMap.OnInfoWindowCloseListener? = null

    fun update() {
        for (infoWindow in infoWindows) {
            infoWindow.update()
        }
    }

    fun setInfoWindowAdapter(infoWindowAdapter: MapLibreMap.InfoWindowAdapter?) {
        this.infoWindowAdapter = infoWindowAdapter
    }

    fun getInfoWindowAdapter(): MapLibreMap.InfoWindowAdapter? = infoWindowAdapter

    fun setAllowConcurrentMultipleOpenInfoWindows(allow: Boolean) {
        allowConcurrentMultipleInfoWindows = allow
    }

    fun isAllowConcurrentMultipleOpenInfoWindows(): Boolean = allowConcurrentMultipleInfoWindows

    fun isInfoWindowValidForMarker(marker: Marker?): Boolean =
        marker != null && (!marker.title.isNullOrEmpty() || !marker.snippet.isNullOrEmpty())

    fun setOnInfoWindowClickListener(listener: MapLibreMap.OnInfoWindowClickListener?) {
        onInfoWindowClickListener = listener
    }

    fun getOnInfoWindowClickListener(): MapLibreMap.OnInfoWindowClickListener? = onInfoWindowClickListener

    fun setOnInfoWindowLongClickListener(listener: MapLibreMap.OnInfoWindowLongClickListener?) {
        onInfoWindowLongClickListener = listener
    }

    fun getOnInfoWindowLongClickListener(): MapLibreMap.OnInfoWindowLongClickListener? = onInfoWindowLongClickListener

    fun setOnInfoWindowCloseListener(listener: MapLibreMap.OnInfoWindowCloseListener?) {
        onInfoWindowCloseListener = listener
    }

    fun getOnInfoWindowCloseListener(): MapLibreMap.OnInfoWindowCloseListener? = onInfoWindowCloseListener

    fun add(infoWindow: InfoWindow) {
        infoWindows.add(infoWindow)
    }
}
